src/Middleware/RateLimiter.php

```diff
@@ -0,0 +1,149 @@
+<?php
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\Middleware;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Psr\SimpleCache\CacheInterface;
+
+/**
+ * Rate limiter limits the number of requests that could be made within a certain period of time
+ */
+final class RateLimiter implements MiddlewareInterface
+{
+    private int $limit = 1000;
+
+    private ?string $cacheKey = null;
+
+    /**
+     * @var callable
+     */
+    private $cacheKeyCallback;
+
+    private int $cacheTtl = 360;
+
+    private CacheInterface $cache;
+
+    private ResponseFactoryInterface $responseFactory;
+
+    private bool $autoincrement = true;
+
+    public function __construct(CacheInterface $cache, ResponseFactoryInterface $responseFactory)
+    {
+        $this->cache = $cache;
+        $this->responseFactory = $responseFactory;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $this->setupCacheParams($request);
+
+        if (!$this->isAllowed()) {
+            return $this->createErrorResponse();
+        }
+
+        if ($this->autoincrement) {
+            $this->increment();
+        }
+
+        return $handler->handle($request);
+    }
+
+    public function setLimit(int $limit): self
+    {
+        $this->limit = $limit;
+
+        return $this;
+    }
+
+    public function setCacheKey(string $key): self
+    {
+        $this->cacheKey = $key;
+
+        return $this;
+    }
+
+    public function setCacheKeyByCallback(callable $callback): self
+    {
+        $this->cacheKeyCallback = $callback;
+
+        return $this;
+    }
+
+    public function setCacheTtl(int $ttl): self
+    {
+        $this->cacheTtl = $ttl;
+
+        return $this;
+    }
+
+    public function setAutoIncrement(bool $increment): self
```

Comment:
What's the idea behind it? Why would one disable increment? Also, there's no way to increment manually except writing to cache (and that's not a good idea).

Reply:
The idea was that you could use this implementation not only to limit requests, but also for example-SMS. 

In this case, you will have to write to the cache manually. 

In my opinion, the obligation of middleware is still to check the limits. 

There was an idea to take out work with a cache in a separate class and there to implement the increment function, but I'm not sure that this is a good idea.


In case of SMS:

```
$middleware = (new RateLimiter($cache, $responseFactory))
    ->setLimit(100)
    ->setInterval(86400)
    ->setCacheKey('send-sms-users');
```

```
 Route::get('/')
      ->prepend($middleware)
      ->to(new ActionCaller(SiteController::class, 'index', $container))
      ->name('site/sms')
```

```
public function sms(CacheInterface $cache, SmsService $smsService): ResponseInterface
{
    if (!$this->user->isGuest()) {
        $smsService->send($this->user);
        $cache->set('send-sms-users', 100);
    }
    ...
}
```

Reply:
Option with a separate class:

RateLimiter:
```
final class RateLimiter implements MiddlewareInterface
{
    private int $limit = 1000;

    private bool $autoincrement = true;

    private RateLimiterCounter $counter;

    private ResponseFactoryInterface $responseFactory;

    public function __construct(RateLimiterCounter $counter, ResponseFactoryInterface $responseFactory)
    {
        $this->counter = $counter;
        $this->responseFactory = $responseFactory;
    }

    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        $this->counter->init($request);

        if (!$this->isAllowed()) {
            return $this->createErrorResponse();
        }

        if ($this->autoincrement) {
            $this->counter->increment();
        }

        return $handler->handle($request);
    }

    public function setLimit(int $limit): self
    {
        $this->limit = $limit;

        return $this;
    }

    public function setAutoIncrement(bool $increment): self
    {
        $this->autoincrement = $increment;

        return $this;
    }

    private function createErrorResponse(): ResponseInterface
    {
        $response = $this->responseFactory->createResponse(429);
        $response->getBody()->write('Too Many Requests');

        return $response;
    }

    private function isAllowed(): bool
    {
        return $this->counter->getCounterValue() < $this->limit;
    }
}
```

RateLimiterCounter:
```
final class RateLimiterCounter
{
    private int $ttl = 360;

    private ?string $key = null;

    /**
     * @var callable
     */
    private $keyCallback;

    private CacheInterface $storage;

    public function __construct(CacheInterface $cache)
    {
        $this->storage = $cache;
    }

    public function init(ServerRequestInterface $request): self
    {
        $this->key = $this->generateKey($request);

        return $this;
    }

    public function setKeyCallback(callable $callback): self
    {
        $this->keyCallback = $callback;

        return $this;
    }

    public function setKey(string $key): self
    {
        $this->key = $key;

        return $this;
    }

    public function setTtl(int $ttl): self
    {
        $this->ttl = $ttl;

        return $this;
    }

    public function increment(): void
    {
        $value = $this->getCounterValue();
        $value++;

        $this->setCounterValue($value);
    }

    public function getCounterValue(): int
    {
        $this->checkOptions();
        return $this->storage->get($this->key, 0);
    }

    private function setCounterValue(int $value): void
    {
        $this->checkOptions();
        $this->storage->set($this->key, $value, $this->ttl);
    }

    private function generateKey(ServerRequestInterface $request): string
    {
        if ($this->keyCallback !== null) {
            return \call_user_func($this->keyCallback, $request);
        }

        return $this->key ?? $this->generateKeyFromRequest($request);
    }

    private function generateKeyFromRequest(ServerRequestInterface $request): string
    {
        return strtolower('rate-limiter-' . $request->getMethod() . '-' . $request->getUri()->getPath());
    }

    private function checkOptions(): void
    {
        if ($this->key === null) {
            throw new \RuntimeException('The cache key cannot be null');
        }
    }
}
```


Usage:
```
        $counter = (new RateLimiterCounter($cache))
           ->setKey('sms-send');

        $middleware = (new RateLimiter($counter, $responseFactory))
            ->setLimit(10)
            ->setAutoIncrement(false);
```

manual increment:
```
        (new RateLimiterCounter($cache))
            ->setKey('sms-send')
            ->increment();
```

Reply:
added counter implementation